internal/saucecloud/cloud.go

```diff
@@ -423,102 +458,78 @@ func checkPathLength(projectFolder string, matcher sauceignore.Matcher) (string,
 	return "", nil
 }
 
-// archiveFolder creates a zip file in the tempDir directory from the given projectFolder.
-func (r *CloudRunner) archiveFolder(project interface{}, tempDir string, projectFolder string, sauceignoreFile string) (string, error) {
-	matcher, err := r.getFileMatcher(projectFolder, sauceignoreFile)
-	if err != nil {
-		return "", err
-	}
+// archiveNodeModules archives node_modules located under rootDir and returns the path to the zip file. Returns an empty
+// string if node_modules doesn't exist or is actively ignored.
+func (r *CloudRunner) archiveNodeModules(tempDir string, rootDir string, matcher sauceignore.Matcher) (string, error) {
+	modDir := filepath.Join(rootDir, "node_modules")
+	ignored := matcher.Match([]string{modDir}, true)
 
-	if path, err := checkPathLength(projectFolder, matcher); err != nil {
-		msg.PathTooLongForArchive(path)
-		return "", err
-	}
-
-	start := time.Now()
+	_, err := os.Stat(modDir)
+	hasMods := err == nil
+	wantMods := len(r.NPMDependencies) > 0
 
-	zipName := filepath.Join(tempDir, "app.zip")
-	z, err := zip.NewFileWriter(zipName, matcher)
-	if err != nil {
-		return "", err
+	if !hasMods && wantMods {
+		return "", fmt.Errorf("unable to access 'node_modules' folder, but you have npm dependencies defined in your project")
 	}
-	defer z.Close()
 
-	rcPath := filepath.Join(tempDir, "sauce-runner.json")
-	if err := jsonio.WriteFile(rcPath, project); err != nil {
-		return "", err
+	if ignored && wantMods {
+		return "", fmt.Errorf("'node_modules' is ignored by sauceignore, but you have npm dependencies defined in your project; please remove 'node_modules' from your sauceignore file")
 	}
 
-	folderContent, err := os.ReadDir(projectFolder)
-	if err != nil {
-		return "", err
-	}
+	var files []string
 
-	totalFileCount := 0
-	for _, child := range folderContent {
-		fileCount, err := z.Add(filepath.Join(projectFolder, child.Name()), "")
-		if err != nil {
-			return "", err
+	// does the user only want a subset of dependencies?
+	if hasMods && wantMods {
+		reqs := node.Requirements(filepath.Join(rootDir, "node_modules"), r.NPMDependencies...)
+		log.Info().Msgf("Found a total of %d related npm dependencies", len(reqs))
+		for _, v := range reqs {
+			files = append(files, filepath.Join("node_modules", v))
 		}
-		totalFileCount += fileCount
-	}
-	fileCount, err := z.Add(rcPath, "")
-	if err != nil {
-		return "", err
 	}
-	totalFileCount += fileCount
 
-	err = z.Close()
-	if err != nil {
-		return "", err
+	// node_modules exists, has not been ignored and a subset has not been specified, so include the entire folder.
+	// This is the legacy behavior (backwards compatible) of saucectl.
+	if hasMods && !ignored && !wantMods {
+		files = append(files, "node_modules")
```

Comment:
This might a good opportunity for a warning since this is usually not the user's intention. Nudge them towards some "best practices".

Reply:
Good point! Let me update the sauce docs as well, so I can include a reference in here.